### ISSUE TITLE
[go_router] GoRouter builds parent route when navigating directly to a subroute (deep link)

### ISSUE THREAD

**Issue by callmephil** [flutter/flutter#141281]
### Steps to reproduce

**The Problem:** 
When using GoRouter to navigate directly to a subroute via a deep link, the parent route's build method is triggered. This results in an unnecessary and costly rebuild of the parent route's screen.

**Steps to Reproduce:**
1. Clone the code sample from [this Gist](https://gist.github.com/callmephil/d5c02c84a788fce17b682db0864074ff).
2. Create a new Flutter project with flutter create go_router_test and replace main.dart with the code from the Gist.
3. don't forget to add below **flutter:** in your dependencies. 
  ```
  flutter_web_plugins:
    sdk: flutter
  ```

4. Run the app in a browser.
5. Open the browser console and clear it.
6. Navigate to a subroute (e.g., www.localhost:PORT/1).
7. Observe that the parent route is built when navigating directly to its child.

Environment:
    (FVM) Flutter Version: 3.16.5
    GoRouter Version: 13.0.1

### Expected results

The parent route should not be built when navigating directly to a subroute.

### Actual results

The parent route is being built, which is unnecessary and causes performance issues.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_web_plugins/url_strategy.dart';
import 'package:go_router/go_router.dart';

final _navKey = GlobalKey<NavigatorState>();

void main() {
  usePathUrlStrategy();

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      routerConfig: GoRouter(
        navigatorKey: _navKey,
        debugLogDiagnostics: true,
        routerNeglect: true,
        initialLocation: '/',
        routes: [
          GoRoute(
            path: '/',
            builder: (context, state) {
              return const ParentScreen();
            },
            routes: [
              GoRoute(
                path: ':id',
                builder: (context, state) {
                  return const ChildScreen();
                },
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class ParentScreen extends StatefulWidget {
  const ParentScreen({super.key});

  @override
  State<ParentScreen> createState() => _ParentScreenState();
}

class _ParentScreenState extends State<ParentScreen> {
  @override
  Widget build(BuildContext context) {
    // This print statement is being called if the child screen is opened.
    print('parent screen has been built');

    return Scaffold(
      appBar: AppBar(
        title: const Text('Parent Screen'),
      ),
      body: Center(
        child: TextButton(
          onPressed: () => context.go('/1'),
          child: const Text('Go to child screen'),
        ),
      ),
    );
  }
}

class ChildScreen extends StatelessWidget {
  const ChildScreen({super.key});

  @override
  Widget build(BuildContext context) {
    print('child screen has been built');
    return Scaffold(
      appBar: AppBar(
        title: const Text('Child Screen'),
      ),
      body: Center(
        child: TextButton(
          onPressed: () => context.go('/'),
          child: const Text('Go to parent screen'),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

![image](https://github.com/flutter/flutter/assets/2213079/ee2043dd-9c22-4ed9-b560-5a65e4f6d2ff)

</details>

### Logs

<details open><summary>Logs</summary>

```console
flutter create go_router_deeplink_test
flutter pub add go_router
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.5, on macOS 14.2.1 23C71 darwin-arm64, locale en-US)
    • Flutter version 3.16.5 on channel stable at /Users/phil/fvm/versions/3.16.5
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (3 weeks ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/phil/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/phil/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C65
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Users/phil/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3.2)
    • IntelliJ at /Users/phil/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin version 233.13135.65

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
Thanks for the detailed report @callmephil 
Can you take a look at [this](https://github.com/flutter/flutter/issues/126785) issue and check if it helps in your case or not ?

**Comment by callmephil** [comment]
> Thanks for the detailed report @callmephil 
> Can you take a look at [this](https://github.com/flutter/flutter/issues/126785) issue and check if it helps in your case or not ?
> 

It's not the same, it's a deep link related bug.

**Comment by darshankawar** [comment]
Thanks for the update. Was able to replicate the reported error.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.16.5, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.16.5 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (13 days ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.19.0-3.0.pre.50, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.19.0-3.0.pre.50 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3e5cad037c (68 minutes ago), 2024-01-10 22:42:23 -0500
    • Engine revision ade9f18fa0
    • Dart version 3.4.0 (build 3.4.0-16.0.dev)
    • DevTools version 2.31.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by chunhtai** [comment]
This may be solve if we set the parent route's maintain state to false, but that will have other consequences. Not sure if that is the right solution

**Comment by 2zqa** [comment]
> [It's not the same, it's a deep link related bug.](https://github.com/flutter/flutter/issues/141281#issuecomment-1886839102)

I still think it's the same issue on a technical level, to load child pages it appears to be necessary to load the parent widgets as well (as per [this comment](https://github.com/flutter/flutter/issues/99245#issuecomment-1090460392)). I think we need some way to determine whether a page is currently in view as a workaround.

I think this issue is a duplicate of
* https://github.com/flutter/flutter/issues/99245

and similar to:
* https://github.com/flutter/flutter/issues/126785
* https://github.com/flutter/flutter/issues/144303

**Comment by japatel225** [comment]
In my case, it is fine that it also builds the parent route. No problem with that.

But just don't want the parent page to be visible (for a second it is visible) when coming via the child deeplink. I tried with `NoTransitionPage`, so the slide-in animation is gone. But the parent page is still visible briefly before the child page appears. Is there a way to achieve that?

**Comment by nicolasbraun** [comment]
**I think this is a more global and critical issue.**

## TL;DR

- Using `context.go` will cause current route to rebuild even if I'm already on the parent of the page i'm going to
- Using `context.push` will cause all routes in the current tree to rebuild
- It also rebuild on pop


## Example

Taking as an example the official named_routes one whose routing definition is

```
- home
- - family
- - - person
```
and simply puting a logs in the builder and build methods

When on `Family` on going to a `Person` , the builder for FamilyRoute get's recalled, causing it to rebuild if not a const Widget.
It's also rebuilds on pop of the `Person` page.

Code https://github.com/nicolasbraun/packages/tree/rebuild_bug_demo

```
[GoRouter] getting location for name: "family", pathParameters: {fid: f1}
[GoRouter] going to /family/f1
flutter: FamilyScreen builder called
flutter: FamilyScreen build
[GoRouter] getting location for name: "person", pathParameters: {fid: f1, pid: p1}, queryParameters: {qid: quid}
[GoRouter] going to /family/f1/person/p1?qid=quid
flutter: PersonScreen builder called
flutter: PersonScreen build
flutter: FamilyScreen builder called
flutter: FamilyScreen build
flutter: FamilyScreen builder called // < poping
flutter: FamilyScreen build

```

https://github.com/user-attachments/assets/9abbd17d-6944-40b6-9305-ddb665078d77

If you move the `PersonRoute`definition outside out the Family one, it does not happen, which makes sense (but in that case the Family screen will not be added to the screen on deeplink)


## But using context.push is even worse

Using .push is quite common to push 'standalone' screens that do no belong the hierachy (shared screen).

If you change the code to use `context.push`  instead of go the route you are calling from  and all its parents are **always** get rebuilt!

For example, moving the personRoute definition outside of the family one 
```
- home
- - family
- - person
```

and calling `context.pushNamed("person"..)` also causes the current FamilyPage to rebuild even if the familyRoute is no longer a parent.
Code : https://github.com/nicolasbraun/packages/tree/rebuild_bug_demo_push_sibling

https://github.com/user-attachments/assets/5cd1966e-fbb2-469b-8f3e-c5744d0b8ba9


Going further and adding a DummyPage, defined at root.
```
- home
- - family
- - - person
- - dummy
```

Calling `context.pushNamed("Dummy.", ..)` from a Person page will cause **both** the Family and the Person page to also rebuild, and rebuild on pop.
 
Code: https://github.com/nicolasbraun/packages/tree/rebuild_bug_demo_push_dumy

https://github.com/user-attachments/assets/bad81b36-3e53-4590-9905-5f5e0f969473

> [!WARNING]  
> This is from my PoV critical, this causes many rebuilds which affect performance and may break apps too. 

Concrete example:
- Imagine an app that uses riverpod for state management 
- There is a userProfile page where the user can choose a picture, this pages is a ConsumerWidget that watches a controller.
- The picture picking page is a route defined at "root", displayed modally, because picture picking is used at other part of the app.

One might want to do:
- click a pick picture button
- `context.push("picturePicking", extra: {"onCallBack: (File file) => ref.read(controlllerProvider).updateUserPicture(file))}`
- ===> breaks, the profile page has been rebuilt, `ref` is no longer avalaible

**Comment by nicolasbraun** [comment]
PS: somehow this does not happen with `ShellRoute` but from my understanding of them, they just use another Navigator. 
It seems overkill to use one just to "wrap" routes, if they don't share common UI builder

**Comment by NoName-gy** [comment]
i don't think it work for shellRoute as well as router that nested inside BottomNavigation.

`
static final rootNavigatorKey = GlobalKey<NavigatorState>();
static final shellNavigatorKey = GlobalKey<NavigatorState>();
...
GoRoute(
                parentNavigatorKey: rootNavigatorKey,
                path: Routes.myCards,
                builder: (context, state) {
                  final viewModel = CardListViewModel(
                    cardRepository: context.read<CardRepository>(),
                  );
                  viewModel.loadCards.execute();
                  return CardListScreen(viewModel: viewModel);
                },
                routes: [
                  GoRoute(
                    parentNavigatorKey: rootNavigatorKey,
                    path: Routes.cardDetail,
                    builder: (context, state) {
                      final card = state.extra as CampaignCard;
                      final viewModel = CardDetailViewModel(
                        cardRepository: context.read<CardRepository>(),
                        card: card,
                      );
                      return CardDetailScreen(viewModel: viewModel);
                    } 
                  ),
                ],
              ),`

**Comment by callmephil** [comment]
@chunhtai can we please reconsider the priority of this issue? I wish I had a way to know the "targeted route" from the router state, so I can simply render a sized box instead of my parent screen if the current route doesn't fully match the target pattern.

**Comment by Wdestroier** [comment]
It's a very weird behavior 😄. I spent almost 2 days figuring out a bug related to a random page redirect caused by a random constructor call caused by GoRouter (no deep link). I found an [answer](https://stackoverflow.com/a/78392723) on SO about why GoRouter builds all parent route widgets.

**Comment by callmephil** [comment]
> It's a very weird behavior 😄. I spent almost 2 days figuring out a bug related to a random page redirect caused by a random constructor call caused by GoRouter (no deep link). I found an [answer](https://stackoverflow.com/a/78392723) on SO about why GoRouter builds all parent route widgets.

Yes, it makes sense that the route needs to be built, but there should be a mechanism that prevents the build of the parent widget while traveling directly to that sub-route.

**Comment by maxencealluin** [comment]
Any follow-up on this issue ? I have noticed the excessive rebuilding and have got rid of most of it simply by adding pagebuilders to my Gorouter routes so far, but it does not work when navigating to a non-child route. This effect is very noticeable in my app which runs much smoother now so it would be great if this was resolved.

**Comment by alperenekin** [comment]
I also have the same problem. The behavior is quite unexpected and it is likely to effect many apps.
